feature/#198 - switched to a simpler (and closer to defaults) color management

monsieurtanuki opened the pull request:
The main change is in `smooth_theme.dart`. Everything comes from there.

Impacted files:
* `choose_page.dart`: simplified
* `home_page.dart`: simplified; used the new method `SmoothTheme.getColor`
* `list_page.dart`: simplified
* `pantry.dart`: used the new method `SmoothTheme.getColor`
* `pantry_button.dart`: used the new method `SmoothTheme.getColor`
* `pantry_dialog_helper.dart`: used the new method `SmoothTheme.getColor`
* `pantry_list_page.dart`
* `pantry_page.dart`: simplified; used the new method `SmoothTheme.getColor`
* `pantry_preview.dart`: used the new method `SmoothTheme.getColor`
* `personalized_ranking_page.dart`: simplified; used the new method `SmoothTheme.getColor`
* `product_list.dart`: used the new method `SmoothTheme.getColor`
* `product_list_button.dart`: used the new method `SmoothTheme.getColor`
* `product_list_dialog_helper.dart`: used the new method `SmoothTheme.getColor`; unrelated uniformisation with pantry list when displaying tinted icons
* `product_list_page.dart`: simplified; used the new method `SmoothTheme.getColor`
* `product_list_preview.dart`: used the new method `SmoothTheme.getColor`
* `product_page.dart`: simplified; used the new method `SmoothTheme.getColor`
* `product_query_page.dart`: no more "white" as default background color
* `profile_page.dart`: simplified
* `smooth_product_card_found.dart`: simplified; no more "white/black" as default (for dark theme reasons)
* `smooth_theme.dart`: create enum `ColorDestination` and a powerful `static Color getColor` method; simplified the colorSchemes to the defaults

M123-dev reviewed (changes requested):
This is a good pr and this will make theoretical color choices easier. However, the code is not the easiest to understand and if I haven't overlooked anything, we use the "standard" colors.
This isn't so beneficial, as we deliberately don't use the material components.
Also there are several things that I don't like about the current choice of colors, but as you said they are only temporary anyway. It would be a lot easier if we have a "manual" overview over the colors, especially in view of the fact that we will adapt the colors often in the future.

But I noticed a few more things.

- The background color of the cards in the scanner are invisible.

- The app bar color from the recent seen products is a different color than the main. I suspect this is wanted, but it is still confusing.

- There is very little contrast in the food comparison screen in White Mode

monsieurtanuki commented:
Thank you @M123-dev for your code review!

You're right: we use the "standard" colors.
Many of the things I've changed are related to a better management of dark/light mode - e.g. not displaying a white card in the dark mode.
I'm not sure what you mean with "This isn't so beneficial, as we deliberately don't use the material components.". Are you thinking about the buttons, for instance?

The only potentially confusing part is with 
```dart
enum ColorDestination {
  APP_BAR_FOREGROUND, // only used in the lists with their colors
  APP_BAR_BACKGROUND, // only used in the lists with their colors
  SURFACE_FOREGROUND,
  SURFACE_BACKGROUND,
  BUTTON_FOREGROUND,
  BUTTON_BACKGROUND,
}
```

But there's no way to make it easy and handle the dark/light modes with several `MaterialColor`s.
In each case, the question is: "instead of the standard color, I want to add a tint of a MaterialColor". If that question doesn't make any sense in the app, you're right, there are many lines of code we could remove.
Let's say that instead of a white background for a card, you want a blue tint.
The code for that is `SmoothTheme.getColor(colorScheme, Colors.blue, ColorDestination.SURFACE_BACKGROUND)`

As for the choice of colors, I didn't choose anything, this is the default in flutter. Of course we'll change the colors. How do you imagine this "manual overview"?

I've fixed this bug that you mentioned: "The background color of the cards in the scanner are invisible."

I don't know what to do with this one: "The app bar color from the recent seen products is a different color than the main. I suspect this is wanted, but it is still confusing." It's definitely intended and I understand this can be confusing. If I remove that, the list color will not be visible when you're on the list page. There's no good or bad choice. It's a choice.

About this one: "There is very little contrast in the food comparison screen in White Mode", if you're talking about the personalized ranking page (between green, grey and red cards), that's correct, because the stronger the card background color is the less contrast you have between the text and the background. On my smartphone I had no problem seeing the differences. Maybe there's something we need to talk about: I assumed that in that case the color should be a light tint, like a card background, but there's the other option to display them more aggressively like buttons with strong backgrounds.

If you want, just for the sake of tests, I can easily switch to an app which will replace all the "confusing" use of `ColorDestination` with the standard colors. Everything will look the same, but that would be a test. What do you think about it?

monsieurtanuki commented:
For the record, when switching off the color management:
![Simulator Screen Shot - iPhone 12 Pro Max - 2021-02-18 at 17 48 43](https://user-images.githubusercontent.com/11576431/108391017-8c5f5300-7211-11eb-89cb-a49396acee11.png)

stephanegigandet reviewed (approved):
The changes make sense to me, especially after reading the comments of the PR.
One thing that would help me understand the code quicker would be to add a one line comment above the functions, to understand what they do. e.g. for getColor : return a specific shade of a color tint, depending on where the color is used and if the current theme is light or dark.

stephanegigandet commented on packages/smooth_app/lib/themes/smooth_theme.dart:
```diff
@@ -1,63 +1,70 @@
     );
   }
 
-  static const ColorScheme _COLOR_LIGHT = ColorScheme.light(
```
Just curious: where are the colors of the different themes defined now? e.g. if I want to add a "light blue polar bear" theme, where would I add specify the colors?

monsieurtanuki replied:
@stephanegigandet What are used now are the default color schemes for light and dark modes.
```dart
const ColorScheme.light({
  this.primary = const Color(0xff6200ee),
  this.primaryVariant = const Color(0xff3700b3),
  this.secondary = const Color(0xff03dac6),
  this.secondaryVariant = const Color(0xff018786),
  this.surface = Colors.white,
  this.background = Colors.white,
  this.error = const Color(0xffb00020),
  this.onPrimary = Colors.white,
  this.onSecondary = Colors.black,
  this.onSurface = Colors.black,
  this.onBackground = Colors.black,
  this.onError = Colors.white,
  this.brightness = Brightness.light,
})

const ColorScheme.dark({
  this.primary = const Color(0xffbb86fc),
  this.primaryVariant = const Color(0xff3700B3),
  this.secondary = const Color(0xff03dac6),
  this.secondaryVariant = const Color(0xff03dac6),
  this.surface = const Color(0xff121212),
  this.background = const Color(0xff121212),
  this.error = const Color(0xffcf6679),
  this.onPrimary = Colors.black,
  this.onSecondary = Colors.black,
  this.onSurface = Colors.white,
  this.onBackground = Colors.white,
  this.onError = Colors.black,
  this.brightness = Brightness.dark,
})
```

If your question is: is it possible to change the primary color, obviously you'll just have to set `primary` and `primaryVariant` (hopefully not `onPrimary`).
If your question is: how can I _add_ color schemes, like "blue polar bear" and "green deer", for the moment we only have one light and one dark. If we want to deal with several animals and color schemes - like in end-user preferences - it's possible but not managed for the moment.

stephanegigandet replied:
Thanks for the precision. I do want to add a blue polar bear and a green deer at some point indeed. :)

M123-dev reviewed (commented):
I lost the PR out of my eyes, sorry for that. 

All the errors have been fixed 👍🏻 

> As for the choice of colors, I didn't choose anything, this is the default in flutter. Of course we'll change the colors. How do you imagine this "manual overview"?

I doubt that we will use the standard colors, so it would be much smarter if we saved the colors somewhere. With "manual overview" I have thought of maybe a object or a list or something similar in which all the colors that we are using are defined. Much like color scheme does, but without the restrictions from there.

monsieurtanuki commented:
@stephanegigandet @M123-dev I've just added a comment for `getColor`, and created #218.
